src/hotspot/share/code/aotCodeCache.cpp

```diff
@@ -926,18 +926,13 @@ AOTCodeEntry* AOTCodeCache::find_code_entry(const methodHandle& method, uint com
   if (is_on() && _cache->cache_buffer() != nullptr) {
     ResourceMark rm;
     const char* target_name = method->name_and_sig_as_C_string();
-    uint hash = java_lang_String::hash_code((const jbyte*)target_name, (int)strlen(target_name));
+    uint hash = (uint)pointer_delta((address)method(), (address)SharedBaseAddress, 1);
     AOTCodeEntry* entry = _cache->find_entry(AOTCodeEntry::Code, hash, comp_level);
     if (entry == nullptr) {
       log_info(aot, codecache, nmethod)("Missing entry for '%s' (comp_level %d, hash: " UINT32_FORMAT_X_0 ")", target_name, (uint)comp_level, hash);
 #ifdef ASSERT
-    } else {
-      uint name_offset = entry->offset() + entry->name_offset();
-      uint name_size   = entry->name_size(); // Includes '/0'
-      const char* name = _cache->cache_buffer() + name_offset;
-      if (strncmp(target_name, name, name_size) != 0) {
-        assert(false, "AOTCodeCache: saved nmethod's name '%s' is different from '%s', hash: " UINT32_FORMAT_X_0, name, target_name, hash);
-      }
+    } else if (method() != entry->method()) {
```

Comment:
As you said we can't use pointers (base could be different).
I suggest to keep original code here.

Reply:
May be add `assert(entry->id() == hash`

Reply:
> As you said we can't use pointers (base could be different).
I suggest to keep original code here.

But we should have already restored entry->method() correctly based on current runtime's base address. So method() should always be same as entry->method().

Reply:
> An other change could be preserving method's name in AOT cache only in debug VM for this compares. For printing we can get name from entry->method()->name_and_sig_as_C_string().

Right, that was in my list as well. I was thinking of not storing the name at all. But I guess we can keep it for debug builds.

Reply:
> > As you said we can't use pointers (base could be different).
> > I suggest to keep original code here.
> 
> But we should have already restored entry->method() correctly based on current runtime's base address. So method() should always be same as entry->method().

Good point. I agree with this then. Please, print full name too in the assert.